File: src/main/java/com/leet/medium/AddTwoNumbers.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	carry := 0
	sentinel := &ListNode{0, nil}
	prev := sentinel

	for l1 != nil && l2 != nil {
		sum := l1.Val + l2.Val + carry
		carry, sum = sum/10, sum%10
		prev.Next = &ListNode{sum, nil}
		prev = prev.Next
		l1, l2 = l1.Next, l2.Next
	}

	for l1 != nil {
		sum := l1.Val + carry
		carry, sum = sum/10, sum%10
		prev.Next = &ListNode{sum, nil}
		prev = prev.Next
		l1 = l1.Next
	}

	for l2 != nil {
		sum := l2.Val + carry
		carry, sum = sum/10, sum%10
		prev.Next = &ListNode{sum, nil}
		prev = prev.Next
		l2 = l2.Next
	}

	if carry != 0 {
		prev.Next = &ListNode{1, nil}
	}

	return sentinel.Next
}

File: src/main/java/com/leet/medium/AddTwoNumbers.exs
# Definition for singly-linked list.

defmodule ListNode do
  @type t :: %__MODULE__{
          val: integer,
          next: ListNode.t() | nil
        }
  defstruct val: 0, next: nil
end

defmodule Solution do
  def add(l1, l2, carry) do
    cond do
      l1 == nil and l2 == nil and carry > 0 ->
        %ListNode{val: carry}

      l1 == nil and l2 == nil ->
        nil

      l1 == nil ->
        sum = l2.val + carry
        %ListNode{val: rem(sum, 10), next: add(l2.next, nil, div(sum, 10))}

      l2 == nil ->
        sum = l1.val + carry
        %ListNode{val: rem(sum, 10), next: add(l1.next, nil, div(sum, 10))}

      true ->
        sum = l1.val + l2.val + carry
        %ListNode{val: rem(sum, 10), next: add(l1.next, l2.next, div(sum, 10))}
    end
  end

  @spec add_two_numbers(l1 :: ListNode.t() | nil, l2 :: ListNode.t() | nil) :: ListNode.t() | nil
  def add_two_numbers(l1, l2) do
    add(l1, l2, 0)
  end
end
